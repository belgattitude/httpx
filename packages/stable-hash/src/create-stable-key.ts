import { isPlainObject } from '@httpx/plain-object';

import { sortArr } from './sort-arr';
import { sortObjKeys } from './sort-obj-keys';

type SupportedDataTypes =
  | Record<string, unknown>
  | unknown[]
  | Date
  | string
  | number
  | boolean
  | bigint
  | null
  | undefined;

type SupportedDataTypesRW = SupportedDataTypes | Readonly<SupportedDataTypes>;

type Options = {
  sortArrayValues?: boolean;
};

/**
 * Create a stable key from a given value useful for caching or memoization.
 *
 * Object keys are sorted to maintain equality between objects with
 * the same keys but in different order.
 *
 * This function is
 * @example
 * ```typescript
 * import { createStableKey } from '@httpx/stable-hash';
 *
 * const params = {
 *   key8: 'a string',
 *   key1: 1,
 *   key3: true,
 *   key2: [3, 2, 1],
 *   key7: {
 *     key2: true,
 *     key1: new Date('2025-02-11T08:58:32.075Z'),
 *   },
 * };
 *
 * const key = createStableKey(params);
 *
 * // Will return a string containing
 * // "{"key1":1,"key2":[1,2,3],"key3":true,"key7":{"key1":"2025-02-11T08:58:32.075Z","key2":true},"key8":"a string"}"
 * ```
 */
export const createStableKey = <T extends SupportedDataTypesRW>(
  value: T,
  options?: Options
): string => {
  const { sortArrayValues = true } = options ?? {};
  return JSON.stringify(value, (_, val) => {
    if (val === undefined) {
      return '[undefined]';
    }
    if (val === null) {
      return null;
    }
    const valType = typeof val;
    if (valType === 'bigint') {
      return `[${val}n]`;
    }
    if (['boolean', 'number', 'string'].includes(valType)) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return val;
    }
    if (sortArrayValues && Array.isArray(val)) {
      return sortArr<unknown>(val);
    }
    if (isPlainObject(val)) {
      return sortObjKeys(val);
    }
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return val;
  });
};
