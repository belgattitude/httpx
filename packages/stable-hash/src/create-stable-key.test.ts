import { createStableKey } from './create-stable-key';

describe('createStableKey', () => {
  describe('base types', () => {
    const stableDate = new Date('2025-02-11T08:58:32.075z');
    const baseTypes = [
      [null, 'null'],
      [undefined, '"[undefined]"'],
      [true, 'true'],
      [false, 'false'],
      [stableDate, '"2025-02-11T08:58:32.075Z"'],
      [[1], '[1]'],
      [[1, 2], '[1,2]'],
      [9_007_199_254_740_991n, '"[9007199254740991n]"'],
    ] as const satisfies [value: unknown, expected: string][];
    it.each(baseTypes)(
      'when %s is provided, should return %s',
      (value, expected) => {
        expect(createStableKey(value)).toBe(expected);
      }
    );
  });

  describe('Object params', () => {
    const baseParams = {
      string: 'string',
      number: 1,
      date: new Date('2025-02-11T08:58:32.075z'),
      boolean: true,
      bigint: BigInt(1),
      null: null,
      undefined: undefined,
    } as const;

    it('should return a sort object params', () => {
      expect(createStableKey(baseParams)).toBe(
        '{"bigint":"[1n]","boolean":true,"date":"2025-02-11T08:58:32.075Z","null":null,"number":1,"string":"string","undefined":"[undefined]"}'
      );
    });

    it('should sort nested objects params', () => {
      const nested = {
        boolean: true,
        nested: { params: baseParams },
      } as const;
      expect(createStableKey(nested)).toBe(
        '{"boolean":true,"nested":{"params":{"bigint":"[1n]","boolean":true,"date":"2025-02-11T08:58:32.075Z","null":null,"number":1,"string":"string","undefined":"[undefined]"}}}'
      );
    });

    it('Test full example', () => {
      const params = {
        key8: 'a string',
        key1: 1,
        key3: true,
        key2: [3, 2, 1],
        key7: {
          key2: true,
          key1: new Date('2025-02-11T08:58:32.075Z'),
        },
      };
      const strHash = createStableKey(params);
      expect(strHash).toMatchSnapshot();
    });
  });
});
