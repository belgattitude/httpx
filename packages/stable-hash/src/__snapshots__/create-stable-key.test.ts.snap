// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createStableKey > Object params > Test full example 1`] = `"{"key1":1,"key2":[1,2,3],"key3":true,"key7":{"key1":"2025-02-11T08:58:32.075Z","key2":true},"key8":"a string"}"`;
